### PR TITLE
Support spot capacity type for ManagedMachinePool

### LINKED ISSUE #2701: Support spot instances for ManagedMachinePool
/kind feature

**Describe the solution you'd like**
`AWSManagedMachinePool` should support spot instances. The SDK provides an option to select the `capacityType` for the node group. We should add the same in CAPA. 

**Anything else you would like to add:**
As a first step we can use the existing `instanceType` field with `capacityType=SPOT`. But the complete implementation should allow selection of multiple instance types.

### PR DESCRIPTION
**Pull request by abhinavnagaraj**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds an option to specify the `capacityType` while creating an EKS node group. 
This enables using spot instances with managed node groups.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2701 

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
EKS: Support spot capacity type for ManagedMachinePool
```

### CODE REVIEW

File: exp/api/v1alpha4/awsmanagedmachinepool_types.go [hunk 1]
```diff
@@ -43,6 +43,16 @@ const (
...
 )
 
+// ManagedMachinePoolCapacityType specifies the capacity type to be used for the managed MachinePool.
+type ManagedMachinePoolCapacityType string
+
+const (
+	// ManagedMachinePoolCapacityTypeOnDemand is the default capacity type, to launch on-demand instances.
+	ManagedMachinePoolCapacityTypeOnDemand ManagedMachinePoolCapacityType = "ON_DEMAND"
+	// ManagedMachinePoolCapacityTypeSpot is the spot instance capacity type to launch spot instances.
+	ManagedMachinePoolCapacityTypeSpot ManagedMachinePoolCapacityType = "SPOT"
```
Comment (randomvariable) [inline exp/api/v1alpha4/awsmanagedmachinepool_types.go:53]: Could we make these enums a bit more Kubernetes friendly,
i.e. `onDemand` and `spot`, and then convert to the AWS (Java) backend constants in the controller?

**Comment by shivi28** [comment]
Hey @abhinavnagaraj Thanks for the PR 🎉 
But as you introduced a new field in v1alpha4, so you need to modify conversion functions.
Feel free to ping me if you need a hand

File: exp/api/v1alpha4/awsmanagedmachinepool_types.go (hunk 1 above)
Comment (abhinavnagaraj) [reply exp/api/v1alpha4/awsmanagedmachinepool_types.go]: Ack

File: exp/api/v1alpha4/awsmanagedmachinepool_types.go [hunk 2]
```diff
@@ -121,6 +131,12 @@ type AWSManagedMachinePoolSpec struct {
 	// machine pool
 	// +optional
 	ProviderIDList []string `json:"providerIDList,omitempty"`
+
+	// CapacityType specifies the capacity type for the ASG behind this pool
+	// +kubebuilder:validation:Enum:=ON_DEMAND;SPOT
+	// +kubebuilder:default:=ON_DEMAND
+	// +optional
+	CapacityType *ManagedMachinePoolCapacityType `json:"capacityType,omitempty"`
```
[the remark this answers is not in the document]
Comment (abhinavnagaraj) [reply exp/api/v1alpha4/awsmanagedmachinepool_types.go:139]: Ack

**Pull request merged**